### ISSUE TITLE
GPU augmentation routing via Kornia in PTL DataModule

### ISSUE THREAD

**Issue by Borda** [roboflow/rf-detr#862]
## Goal

Add automatic GPU-side augmentation as an opt-in alternative to the current CPU-based Albumentations
pipeline. When `augmentation_backend="auto"` (or `"gpu"`), the `RFDETRDataModule` applies
[Kornia](https://kornia.readthedocs.io) transforms inside `on_after_batch_transfer` — **after** the
batch is already resident on the GPU — instead of running Albumentations per-sample in
`Dataset.__getitem__`. The CPU path remains unchanged and is the default.

**Phase 1 scope**: detection only. Segmentation mask augmentation is deferred to phase 2.

The PTL hook `on_after_batch_transfer(self, batch, dataloader_idx)` runs after
`transfer_batch_to_device`. It receives the batch as tensors already on the target device, making it
the canonical insertion point for GPU-native augmentation.

## How PTL DataModule enables this

```python
class RFDETRDataModule(LightningDataModule):
    def on_after_batch_transfer(self, batch, dataloader_idx):
        # batch is (NestedTensor, list[dict targets]) — already on device
        if self.trainer.training and self._kornia_pipeline is not None:
            samples, targets = batch
            img = samples.tensors           # [B, C, H, W] — float32, NOT yet normalized
            boxes_padded, valid = collate_boxes(targets, img.device)
            img_aug, boxes_aug = self._kornia_pipeline(img, boxes_padded)
            img_aug = self._normalize(img_aug)          # Normalize after augmentation
            targets = unpack_boxes(boxes_aug, valid, targets, *img_aug.shape[-2:])
            batch = (NestedTensor(img_aug, samples.mask), targets)
        return batch
```

**Critical design constraint (confirmed by sw-engineer review)**: Images and boxes at
`on_after_batch_transfer` time must be **pre-normalization**. When GPU backend is active, the CPU
pipeline must stop before `Normalize` — augmentation happens first (GPU), then Normalize (GPU).

## Affected files (amended after 3-agent review)

- `pyproject.toml` — add `kornia>=0.7,<1` under new `[project.optional-dependencies]` group `kornia`
- `src/rfdetr/config.py` — add `TrainConfig.augmentation_backend: Literal["cpu", "auto", "gpu"] = "cpu"`
- `src/rfdetr/datasets/coco.py` — add `apply_cpu_augmentation: bool = True` to `make_coco_transforms` and `make_coco_transforms_square_div_64`; wire through `build_coco`, `build_roboflow_from_coco`
- `src/rfdetr/datasets/yolo.py` — wire `apply_cpu_augmentation` through `build_roboflow_from_yolo`
- `src/rfdetr/datasets/kornia_transforms.py` — **new file**: `build_kornia_pipeline`, `build_normalize`, `collate_boxes`, `unpack_boxes`
- `src/rfdetr/datasets/aug_config.py` — note Kornia-supported keys per preset
- `src/rfdetr/_namespace.py` — verify `augmentation_backend` auto-forwards to namespace (it will by default); add to `_TC_NON_NAMESPACE_FIELDS` only if build functions should NOT read it
- `src/rfdetr/training/module_data.py` — add `on_after_batch_transfer`; build Kornia pipeline + normalize in `setup("fit")`; resolve backend
- `tests/training/test_module_data.py` — extend with `on_after_batch_transfer` + backend resolution tests
- `tests/datasets/test_kornia_transforms.py` — **new file**: pipeline builder + bbox utility tests

> **Note on `o365.py`**: `build_o365_raw` calls `make_coco_transforms` but does not pass `aug_config`
> and is not a primary training path. O365 + GPU augmentation is out of scope for Phase 1; add a
> guard/warning if `augmentation_backend != "cpu"` and dataset is O365.

## `augmentation_backend` values

| Value | Behaviour |
|---|---|
| `"cpu"` | Always Albumentations on CPU. Default. Explicit opt-out even when GPU is available. |
| `"auto"` | Kornia if CUDA available **and** `kornia` installed. Falls back to CPU with `WARNING` — no error. |
| `"gpu"` | Force Kornia. `RuntimeError` if no CUDA. `ImportError` (with install hint) if kornia missing. |

`"auto"` is the recommended setting for GPU users. `"gpu"` is the loud-failure mode for scripts.

## Risks and mitigations

| # | Risk | Status |
|---|---|---|
| 1 | Parameter mapping completeness | ✅ Resolved |
| 2 | Bounding-box handling | ✅ Resolved — utility functions + `bbox_xyxy` data key |
| 3 | Instance mask support | ⚠️ Acceptable — deferred to phase 2 |
| 4 | Pipeline split fragility | ✅ Resolved — `apply_cpu_augmentation` flag + Normalize split |
| 5 | Default / opt-out behaviour | ✅ Resolved |
| 6 | Dependency weight | ✅ Resolved |
| 7 | CI coverage gap | ⚠️ Mitigated — see test plan below |

---

### Risk 1 ✅ — Parameter mapping completeness

**Actual transforms in all presets** (complete set from `aug_config.py`):

| Albumentations key | Preset(s) | Params used |
|---|---|---|
| `HorizontalFlip` | all | `p` |
| `VerticalFlip` | aggressive, aerial | `p` |
| `Rotate` | aggressive, aerial | `limit` (scalar **or** tuple), `p` |
| `Affine` | aggressive | `scale`, `translate_percent`, `rotate`, `shear`, `p` |
| `ColorJitter` | aggressive | `brightness`, `contrast`, `saturation`, `hue`, `p` |
| `RandomBrightnessContrast` | conservative, aerial, industrial | `brightness_limit`, `contrast_limit`, `p` |
| `GaussianBlur` | industrial | `blur_limit` (int), `p` |
| `GaussNoise` | industrial | `std_range` (tuple), `p` |

**Mitigation**: `_REGISTRY` dict mapping each key to a factory `(params_dict) -> K.AugmentationBase2D`.
Raise `ValueError` upfront for unknown keys. See parameter translation table below.

### Risk 2 ✅ — Bounding-box handling (detection only, phase 1)

Kornia `AugmentationSequential(data_keys=["input", "bbox_xyxy"])` handles bbox transformation
automatically. It expects `[B, N_max, 4]` absolute pixel xyxy coordinates.

**Important**: boxes must be in xyxy absolute pixel format when passed to Kornia. This is correct
**only if Normalize has not yet run** (see pipeline split below — Normalize is stripped from CPU
path when GPU backend is active, so boxes remain in xyxy absolute format at hook time).

Two utility functions in `kornia_transforms.py`, both with unit tests:

```python
def collate_boxes(targets: list[dict], device: torch.device) -> tuple[Tensor, Tensor]:
    """Pack variable-length boxes into [B, N_max, 4] with zero padding.
    Returns (boxes_padded [B, N_max, 4], valid_mask [B, N_max] bool).
    Handles: B=0, N=0 per image, N_max=0."""

def unpack_boxes(boxes_aug: Tensor, valid: Tensor, targets: list[dict], H: int, W: int) -> list[dict]:
    """Clamp [0,W]×[0,H], remove zero-area boxes, write back into targets.
    Removes corresponding labels/area/iscrowd entries for removed boxes.
    Handles: all boxes removed (empty output), partial removal."""
```

Phase 1: if `model_config.segmentation_head` is `True`, `on_after_batch_transfer` skips
augmentation entirely and logs one warning. Using the config flag (not runtime mask key inspection)
is more reliable — it is set before training starts and never changes.

### Risk 3 ⚠️ — Instance mask support (deferred to phase 2)

Deferred. Kornia supports `List[Tensor]` for the `"mask"` data key (one `[N_i, H, W]` tensor per
image), which maps directly to the existing target dict format. Phase 2 adds this path.

### Risk 4 ✅ — Pipeline split (Normalize must move to GPU path)

**This is more than a single boolean flag.** When GPU backend is active, the CPU pipeline must
stop before BOTH `aug_wrappers` AND `Normalize`. Two separate concerns:

1. `apply_cpu_augmentation: bool = True` — omits `aug_wrappers` (Albumentations). Same as before.
2. `apply_normalize: bool = True` — omits `Normalize` from the CPU pipeline. New requirement.

Or combine into one flag: `gpu_postprocess: bool = False`. When `True`: no aug_wrappers, no
Normalize. Simpler to thread through the call chain.

```python
def make_coco_transforms(..., gpu_postprocess: bool = False) -> Compose:
    if image_set == "train":
        pipeline = [*resize_wrappers]
        if not gpu_postprocess:
            pipeline += [*aug_wrappers]
        pipeline += [to_image, to_float]
        if not gpu_postprocess:
            pipeline += [normalize]
        return Compose(pipeline)
```

`build_kornia_pipeline` then returns both the augmentation pipeline and a `build_normalize()`
utility that creates the GPU-side Normalize transform. The DataModule applies them in sequence in
`on_after_batch_transfer`.

**NestedTensor reconstruction**: After Kornia augmentation on `samples.tensors` (shape `[B, C, H, W]`),
rebuild with `NestedTensor(augmented_tensors, samples.mask)`. The `.mask` (True = padding) stays
valid: geometric transforms on zero-padded regions produce near-zero values, which is harmless.
Kornia operates on the entire padded tensor uniformly.

### Risk 5 ✅ — Default / opt-out behaviour

- Default `"cpu"` — zero behaviour change for existing users.
- All backend resolution decisions logged at `INFO` level.
- `setup()` idempotency: build `_kornia_pipeline` only once (guard: `if self._kornia_pipeline is None`), consistent with the existing dataset build guard pattern in the DataModule.

### Risk 6 ✅ — Dependency weight

Lazy import. Extra group named `kornia` (not `gpu-aug`) to match convention (`lora`, `onnx`, `trt`):

```python
def _require_kornia():
    try:
        import kornia.augmentation  # noqa: F401
    except ImportError as e:
        raise ImportError(
            "GPU augmentation requires kornia. "
            "Install with: pip install 'rfdetr[kornia]'"
        ) from e
```

| Scenario | Outcome |
|---|---|
| `auto` + kornia not installed | Fall back to CPU, log `WARNING` |
| `auto` + no CUDA | Fall back to CPU, log `INFO` |
| `gpu` + kornia not installed | `ImportError` with install hint |
| `gpu` + no CUDA | `RuntimeError("augmentation_backend='gpu' requires a CUDA device")` |

### Risk 7 ⚠️ — CI coverage

**Mock strategy for `self.trainer.training`** (required for unit tests):
```python
# In test — mirrors existing pattern from tests/training/conftest.py
mock_trainer = MagicMock(training=True)
type(dm).trainer = property(lambda self: mock_trainer)
# Add autouse cleanup fixture to tests/training/conftest.py:
#   _restore_rfdetr_datamodule_trainer_property  (mirrors existing _restore_rfdetr_module_trainer_property)
```

**Test matrix — CPU CI** (no GPU required; Kornia runs on CPU tensors):

| Test class | Cases |
|---|---|
| `TestBuildKorniaPipeline` | Each preset config; unknown key `ValueError`; empty config `{}`; known + unknown mix |
| `TestCollateBoxes` | Normal batch; B=0; N=0 per image; single image B=1; all-zero-area after aug |
| `TestUnpackBoxes` | All boxes clipped to zero-area; partial removal; `labels`/`area`/`iscrowd` sync |
| `TestRotateFactory` | `limit` as scalar; `limit` as tuple `(90, 90)` (AUG_AERIAL) |
| `TestOnAfterBatchTransfer` | Training=True applies aug; training=False skips; segmentation model skips |
| `TestBackendResolution` | `auto`+no-CUDA→CPU (mock `torch.cuda.is_available`); `auto`+no-kornia→CPU (mock import); `gpu`+no-CUDA→RuntimeError; `gpu`+no-kornia→ImportError |
| `TestGpuPostprocessFlag` | `make_coco_transforms("train", ..., gpu_postprocess=True)` omits aug + normalize |

**GPU CI** (`@pytest.mark.gpu`):
- `test_on_after_batch_transfer_gpu` with `augmentation_backend="gpu"` on real CUDA batch

**CPU CI parameterisation**: tests parameterise over `["cpu", "auto"]` only. The `"gpu"` happy path
is `@pytest.mark.gpu`. Error paths for `"gpu"` (no CUDA, no kornia) run in CPU CI as standalone
non-parametrised tests in `TestBackendResolution`.

## Parameter translation table (`kornia_transforms.py` implementation reference)

| Albumentations | Kornia class | Parameter conversion |
|---|---|---|
| `HorizontalFlip(p)` | `K.RandomHorizontalFlip(p=p)` | Direct |
| `VerticalFlip(p)` | `K.RandomVerticalFlip(p=p)` | Direct |
| `Rotate(limit, p)` | `K.RandomRotation(degrees=limit if isinstance(limit, tuple) else (-limit, limit), p=p)` | `limit` may be scalar OR tuple — handle both |
| `Affine(scale, translate_percent, rotate, shear, p)` | `K.RandomAffine(degrees=rotate, translate=translate_percent, scale=scale, shear=shear, p=p)` | `translate_percent` is already a fraction tuple |
| `ColorJitter(brightness, contrast, saturation, hue, p)` | `K.ColorJiggle(brightness=brightness, contrast=contrast, saturation=saturation, hue=hue, p=p)` | Direct — same multiplicative semantics |
| `RandomBrightnessContrast(brightness_limit, contrast_limit, p)` | `K.ColorJiggle(brightness=brightness_limit, contrast=contrast_limit, p=p)` | Kornia interprets single float as `(1-x, 1+x)` internally — pass `brightness_limit` directly |
| `GaussianBlur(blur_limit, p)` | `K.RandomGaussianBlur(kernel_size=(3, blur_limit if blur_limit % 2 == 1 else blur_limit+1), sigma=(0.1, 2.0), p=p)` | Round `blur_limit` up to odd; sigma hardcoded (no Albumentations equivalent) |
| `GaussNoise(std_range, p)` | `K.RandomGaussianNoise(std=std_range[1], p=p)` | Upper bound of `std_range`; Kornia takes fixed std (acceptable approximation — document) |

> **Note on Kornia naming**: `K.ColorJiggle` exists in Kornia 0.7.x; `K.ColorJitter` was added as
> an alias in later versions. Use `K.ColorJiggle` for `>=0.7` compatibility.

## `apply_augmentation` propagation chain (amended)

```
TrainConfig.augmentation_backend
    │  (auto-forwarded to namespace by _namespace_from_configs — no _TC_NON_NAMESPACE_FIELDS change needed)
    ▼
build_dataset(image_set, ns, resolution)        [datasets/__init__.py]
    │
    ├─► build_coco(args, ...)                   [datasets/coco.py]
    │       gpu_postprocess = (args.augmentation_backend != "cpu")
    │       make_coco_transforms(..., gpu_postprocess=gpu_postprocess)
    │
    ├─► build_roboflow_from_coco(args, ...)     [datasets/coco.py]
    │       (same pattern)
    │
    └─► build_roboflow_from_yolo(args, ...)     [datasets/yolo.py]
            (same pattern)

build_o365_raw: out of scope Phase 1 — guard with warning if augmentation_backend != "cpu"
```

## Suggested implementation order (amended)

1. `kornia_transforms.py` — `build_kornia_pipeline` + `build_normalize` + `collate_boxes` + `unpack_boxes` + unit tests (including edge cases)
2. `config.py` — add `augmentation_backend` field (`Literal["cpu", "auto", "gpu"]`, default `"cpu"`)
3. `coco.py` — add `gpu_postprocess: bool = False` to `make_coco_transforms` and `make_coco_transforms_square_div_64`; wire through `build_coco`, `build_roboflow_from_coco`
4. `yolo.py` — wire `gpu_postprocess` through `build_roboflow_from_yolo`
5. `module_data.py` — backend resolution in `setup("fit")` (with idempotency guard) + `on_after_batch_transfer` hook; use `model_config.segmentation_head` for phase 1 skip
6. `_namespace.py` — verify `augmentation_backend` auto-forwards (confirm no action needed)
7. `tests/training/conftest.py` — add `_restore_rfdetr_datamodule_trainer_property` autouse fixture
8. Integration + backend resolution tests in `test_module_data.py`; `aug_config.py` docstring

## Follow-up command

```
/develop feature add GPU augmentation routing via Kornia to RFDETRDataModule on_after_batch_transfer
```

**Comment by robmarkcole** [comment]
Also consider that Albumentations moved to a commercial license. TorchGeo uses Kornia, and are investigating torchvision due to compatibility with model exports https://github.com/torchgeo/torchgeo/issues/3108

**Comment by edgarriba** [comment]
apart from image what transforms you need, only the boxes? i'm experimenting with a new rust based augmentations pipeline that benchmarks quite positive again know libraries and gil free or multithreaded pipeline e.g with Ray data. One more goal we have besides helping with training, is whether inference can run efficiently in embedded/cpuvia onnxruntime or rust based solutions like candle/burnt
